File: synpay-dashboard-frontend/api/roles.ts
/**
 * Role Management API Client
 *
 * Calls the FastAPI gateway at /api/roles, which forwards
 * all requests to Spring Boot Integration Core.
 *
 * This module does NOT enforce RBAC — that's handled by Spring Boot.
 */

import { authHeader } from '@/lib/auth'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

// ── API Response Types (mirror Spring Boot DTOs) ─────────────────

export interface ApiEnvelope<T> {
  success: boolean
  message?: string
  data?: T
}

export interface ApiPermissionItem {
  permissionId: number
  key: string
  name: string
  enabled: boolean
}

/** Returned by both list and detail endpoints */
export interface ApiRoleResponse {
  roleId: number
  code: string
  name: string
  description: string | null
  responsibility: string | null
  createdAt: string | null
  updatedAt: string | null
  userCount: number
  /** null in list, populated in detail */
  permissions: ApiPermissionItem[] | null
}

// ── Request Types ────────────────────────────────────────────────

export interface CreateRolePayload {
  code: string
  name: string
  description?: string
  responsibility?: string
}

export interface UpdateRolePayload {
  name?: string
  description?: string
  responsibility?: string
}

export interface PermissionAssignment {
  permissionId: number
  enabled: boolean
}

export interface AssignPermissionsPayload {
  permissions: PermissionAssignment[]
}

// ── API Error ────────────────────────────────────────────────────

export class ApiError extends Error {
  status: number
  constructor(message: string, status: number) {
    super(message)
    this.name = 'ApiError'
    this.status = status
  }
}

// ── Helper ───────────────────────────────────────────────────────

async function request<T>(url: string, options: RequestInit = {}): Promise<T> {
  const response = await fetch(`${API_BASE}${url}`, {
    headers: {
      'Content-Type': 'application/json',
      ...authHeader(),
      ...options.headers,
    },
    ...options,
  })

  const body: ApiEnvelope<T> = await response.json()

  if (!response.ok || !body.success) {
    throw new ApiError(
      body.message ?? `Request failed with status ${response.status}`,
      response.status,
    )
  }

  return body.data as T
}

// ── API Functions ────────────────────────────────────────────────

/** List all roles with user counts. Requires role.read permission. */
export async function fetchRoles(): Promise<ApiRoleResponse[]> {
  return request<ApiRoleResponse[]>('/api/roles')
}

/** Get role detail with permissions. Requires role.read permission. */
export async function fetchRole(id: number): Promise<ApiRoleResponse> {
  return request<ApiRoleResponse>(`/api/roles/${id}`)
}

/** Create a new role. Requires role.write permission. */
export async function createRole(payload: CreateRolePayload): Promise<ApiRoleResponse> {
  return request<ApiRoleResponse>('/api/roles', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

/** Update an existing role. Requires role.write permission. */
export async function updateRole(id: number, payload: UpdateRolePayload): Promise<ApiRoleResponse> {
  return request<ApiRoleResponse>(`/api/roles/${id}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

/** Assign or update permissions for a role. Requires permission.assign permission. */
export async function assignPermissions(
  roleId: number,
  payload: AssignPermissionsPayload,
): Promise<ApiRoleResponse> {
  return request<ApiRoleResponse>(`/api/roles/${roleId}/permissions`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

File: synpay-dashboard-python-api/routes/roles.py
"""
Role management API routes.

These endpoints are consumed by the frontend (Next.js) and forward
all requests to the Spring Boot Integration Core for RBAC enforcement,
business logic, and audit logging.
"""

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from dependencies import extract_token, get_client_ip
from spring_client import forward_request

router = APIRouter(prefix="/api/roles", tags=["Role Management"])


# ── Request schemas (mirror Spring Boot DTOs) ────────────────────

class CreateRoleBody(BaseModel):
    code: str = Field(..., min_length=1, max_length=80)
    name: str = Field(..., min_length=1, max_length=150)
    description: str | None = None
    responsibility: str | None = None


class UpdateRoleBody(BaseModel):
    name: str | None = Field(None, max_length=150)
    description: str | None = None
    responsibility: str | None = None


class PermissionEntry(BaseModel):
    permission_id: int = Field(..., alias="permissionId")
    enabled: bool

    model_config = {"populate_by_name": True}


class AssignPermissionsBody(BaseModel):
    permissions: list[PermissionEntry] = Field(..., min_length=1)


# ── Helpers ───────────────────────────────────────────────────────

def _spring_response(response) -> JSONResponse:
    """Convert an httpx response into a FastAPI JSONResponse."""
    return JSONResponse(
        status_code=response.status_code,
        content=response.json(),
    )


# ── GET /api/roles ───────────────────────────────────────────────

@router.get("")
async def list_roles(
    request: Request,
    token: str = Depends(extract_token),
):
    """
    List all roles with user counts.
    Forwards to: GET /internal/roles
    """
    response = await forward_request(
        "GET",
        "/internal/roles",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── GET /api/roles/{role_id} ─────────────────────────────────────

@router.get("/{role_id}")
async def get_role(
    role_id: int,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Get detailed information for a specific role including permissions.
    Forwards to: GET /internal/roles/{id}
    """
    response = await forward_request(
        "GET",
        f"/internal/roles/{role_id}",
        token,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── POST /api/roles ──────────────────────────────────────────────

@router.post("", status_code=201)
async def create_role(
    body: CreateRoleBody,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Create a new role.
    Forwards to: POST /internal/roles
    """
    payload: dict = {
        "code": body.code,
        "name": body.name,
    }
    if body.description is not None:
        payload["description"] = body.description
    if body.responsibility is not None:
        payload["responsibility"] = body.responsibility

    response = await forward_request(
        "POST",
        "/internal/roles",
        token,
        json_body=payload,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── PUT /api/roles/{role_id} ─────────────────────────────────────

@router.put("/{role_id}")
async def update_role(
    role_id: int,
    body: UpdateRoleBody,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Update an existing role.
    Forwards to: PUT /internal/roles/{id}
    """
    payload: dict = {}
    if body.name is not None:
        payload["name"] = body.name
    if body.description is not None:
        payload["description"] = body.description
    if body.responsibility is not None:
        payload["responsibility"] = body.responsibility

    response = await forward_request(
        "PUT",
        f"/internal/roles/{role_id}",
        token,
        json_body=payload,
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)


# ── PUT /api/roles/{role_id}/permissions ─────────────────────────

@router.put("/{role_id}/permissions")
async def assign_permissions(
    role_id: int,
    body: AssignPermissionsBody,
    request: Request,
    token: str = Depends(extract_token),
):
    """
    Assign or update permissions for a role.
    Forwards to: PUT /internal/roles/{id}/permissions
    """
    response = await forward_request(
        "PUT",
        f"/internal/roles/{role_id}/permissions",
        token,
        json_body={
            "permissions": [
                {"permissionId": p.permission_id, "enabled": p.enabled}
                for p in body.permissions
            ],
        },
        forwarded_for=get_client_ip(request),
        user_agent=request.headers.get("User-Agent"),
    )
    return _spring_response(response)
